MKLDNN elementwis_add with default broadcast operations

tpatejko opened the pull request:
This PR implements `elementwise_add` operator based on MKLDNN sum primitive with a fallback to default functors when tensors' dimensions are of different sizes.

tpatejko commented:
@luotao1 Could you have a look at this PR or assign more reviewers?

@pzelazko-intel Could you have a look once again at the changes, especially regarding MKLDNN format changes?

pzelazko-intel commented on paddle/fluid/framework/data_layout_transform.cc:
```diff
@@ -148,9 +148,12 @@ void TransDataLayoutFromMKLDNN(const OpKernelType& kernel_type_for_var,
 
   memory::format in_format =
-      in_tz.size() == 2 ? memory::format::nc : in.format();
+      in_tz.size() == 1 ? memory::format::x
```
I think you can consider adding function taking size and format as parameters. This block of code is repeated 3 times.

pzelazko-intel commented on paddle/fluid/operators/elementwise_op.h:
```diff
@@ -40,6 +44,24 @@ class ElementwiseOp : public framework::OperatorWithKernel {
+      return framework::OpKernelType(input_data_type, ctx.GetPlace(),
+                                     framework::DataLayout::kMKLDNN,
+                                     framework::LibraryType::kMKLDNN);
+    } else {
```
I think you can put #endif here instead of else for simplification.

pzelazko-intel commented on paddle/fluid/operators/elementwise_op.h:
```diff
@@ -138,6 +162,24 @@ class ElementwiseOpGrad : public framework::OperatorWithKernel {
+      return framework::OpKernelType(input_data_type, ctx.GetPlace(),
+                                     framework::DataLayout::kMKLDNN,
+                                     framework::LibraryType::kMKLDNN);
+    } else {
```
same here

luotao1 commented on python/paddle/fluid/tests/unittests/test_elementwise_add_mkldnn_op.py:
```diff
@@ -0,0 +1,130 @@
+        self.use_mkldnn = True
+
+
+class TestMKLDNNElementwiseAddOp_scalar(TestMKLDNNElementwiseAddOp):
```
if `class TestMKLDNNElementwiseAddOp_scalar(TestElementwiseAddOp_scalar)`, 
then `init_input_output ` could be removed. And only set `def init_kernel_type` is OK
```
class TestMKLDNNElementwiseAddOp_scalar(TestElementwiseAddOp_scalar):
    def init_kernel_type(self):
        self.use_mkldnn = True
```

tpatejko commented:
@luotao1 I incorporated changes you mentioned. However, MKLDNN does not work with data of number of dimensions 3, so I changed the sizes of tensors wherever it was necessary.

tpatejko commented:
@luotao1 @tensor-tang Could you have a look at this PR?

tensor-tang commented on paddle/fluid/operators/elementwise_add_mkldnn_op.cc:
```diff
@@ -0,0 +1,190 @@
+                     "Wrong layout/format set for X tensor");
+      PADDLE_ENFORCE(y->layout() == DataLayout::kMKLDNN &&
+                         y->format() != memory::format::format_undef,
+                     "Wrong layout/format set for X tensor");
```
typo `Y tensor`
